sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs

```diff
@@ -184,19 +184,21 @@ private ProgramState ProcessStreamDisposingTypes(ProgramState programState, Synt
 
                 foreach (var objectCreationExpression in objectCreationExpressions)
                 {
-                    var objectType = this.semanticModel.GetSymbolInfo(objectCreationExpression.Type)
-                    .Symbol
-                    .GetSymbolType();
-
-                    if (objectType.DerivesOrImplementsAny(typesDisposingUnderlyingStream) &&
-                        objectCreationExpression.ArgumentList?.Arguments is IEnumerable<ArgumentSyntax> argumentList)
+                    if (this.semanticModel.GetSymbolInfo(objectCreationExpression.Type).Symbol is ISymbol symbol &&
+                        symbol.GetSymbolType() is ITypeSymbol objectType &&
+                        objectType.DerivesOrImplementsAny(typesDisposingUnderlyingStreamWithLeaveOpenArgumentIndex.Keys.ToImmutableArray()) &&
+                        objectCreationExpression.ArgumentList?.Arguments is IEnumerable<ArgumentSyntax> argumentList &&
+                        argumentList.Any())
                     {
                         // Checks for the 'leaveOpen' argument. It is at position 4 if present for StreamReader constructor, while it is at position 3 for StreamWriter.
                         // See #2491 for more details
-                        var leaveOpenArgumentIndex = objectType.DerivesOrImplements(KnownType.System_IO_StreamReader) ? 4 : 3;
-                        if (!HasLeaveOpenArgument(argumentList, leaveOpenArgumentIndex))
+                        var leaveOpenArgumentIndex = typesDisposingUnderlyingStreamWithLeaveOpenArgumentIndex
+                            .Where(entry => objectType.DerivesOrImplements(entry.Key))
+                            .Select(entry => entry.Value)
+                            .First();

```

Comment:
this will throw `InvalidOperationException` if the list is empty - should it be `FirstOrDefault`?

Reply:
This should always work, because of the outer if-condition `objectType.DerivesOrImplementsAny(typesDisposingUnderlyingStreamWithLeaveOpenArgumentIndex.Keys.ToImmutableArray())`. It implies this `Where` query will always have (at least, although in this case it is exactly) one element